Rb3gen2: Add support for Open Boot firmware (TF-A, OP-TEE and U-Boot) build

b49020 opened the pull request:
Currently the open boot firmware stack has only been enabled on RB3Gen2 platform which can be built using following:

```
$ kas build ci/rb3gen2-core-kit-open-fw.yml:ci/meta-arm.yml
```

And currently the boot stack includes TF-A, OP-TEE and U-Boot. In future the plan is to enable upstream edk2 as well. Along with that the next target for open boot firmware is Lemans based IoT EVK platform.

Right now the build generates 2 firmware payloads as bl2.elf (unsigned) and fip.elf (test signed using qtestsign). It is required to sign bl2.elf with QTI signature using sectools but in future the plan is to drop QTI signature requirement with an updated release of XBL/XBL-SEC.

Once signing is done, one need to update following binaries in qcomflash tarball and then proceed with QDL flashing:
- tz.mbn -> bl2.mbn
- uefi.elf -> fip.elf

koenkooi commented:
> Any ideas how can I fix the yocto run checks here? I have already added `meta-arm` as dependency for `meta-qcom` and added `meta-arm` repo to be part of repos in the KAS `base.yml` file. However, the build still says:
> 
> ```
> ERROR: Layer meta-qcom depends on meta-arm and isn't found.
> ```

I'm not sure if ordering matters in KAS, but could you try changing the repo section to this?
```yaml
  meta-arm:
    url: https://git.yoctoproject.org/git/meta-arm
    layers:
      meta-arm-toolchain:
      meta-arm:
```

That's the only thing that stood out to me as a difference from a working config with meta-arm included.

ndechesne commented on ci/open-boot-firmware.yml:
```diff
@@ -0,0 +1,12 @@
+
+local_conf_header:
+  firmware: |
+    EXTRA_IMAGEDEPENDS += "trusted-firmware-a"
```
these configs seem to qualify to be a MACHINE FEATURE, so that they can be set per machine instead of globally. and then we can define an 'open' RB3g2?

quaresmajose commented on recipes-security/optee/optee-os-tadevkit_%.bbappend:
Same here

b49020 commented on recipes-security/optee/optee-test-qcm6490.inc:
```diff
@@ -0,0 +1,3 @@
+# qcm6490 specific configuration
+
+COMPATIBLE_MACHINE = "qcm6490"
```
`meta-arm` does enforce the `COMPATIBLE_MACHINE` here since all the TAs being built in `optee-test` have a dependency on `optee-os`. It will fail otherwise to build.

ricardosalveti replied:
Should have a dependency on optee-os-tadevkit, which should also be generic, but I see that everything ended up being machine specific in meta-arm, so we can keep this way indeed.

b49020 commented on ci/open-boot-firmware.yml:
```diff
@@ -0,0 +1,12 @@
+    PREFERRED_PROVIDER_virtual/bootloader = "u-boot"
+    TFA_UBOOT:pn-trusted-firmware-a = "1"
+  optee: |
+    IMAGE_INSTALL:append = " optee-test optee-client optee-os-ta"
```
Sure, I can do that.

b49020 commented:
PR updated to incorporate review comments.

b49020 commented:
Although I have fixed the CI issues which were related to this PR however the current CI failure:

```
Parsing recipes...ERROR: /work/oe-core/meta/recipes-kernel/linux/linux-yocto_6.17.bb: Unable to get checksum for linux-yocto SRC_URI entry qcom.scc: file could not be found
```

doesn't seem to be related to this PR, any ideas?

b49020 commented:
The CI is all green now, a gentle ping for any further comments here.

ndechesne commented on ci/open-boot-firmware.yml:
```diff
@@ -0,0 +1,8 @@
+
+local_conf_header:
+  firmware: |
+    MACHINE_FEATURES:qcm6490:append = " tfa optee u-boot"
```
I don't like changing the MACHINE config in the KAS config file. I would rather have a difference machine, what do you think?

lumag replied:
> I don't know if it's a functional difference or not.. but we usually put the :append or :prepend first in the list of overrides.

It is: `foo:soc:append` appends to `foo:soc`, which then override full `foo` value. `foo:append:soc`: makes `foo:append` work only on `soc`, which then appends to the existing `foo` variable without redefining it completely.

I also don't like the idea of changing MACHINE_FEATURES here. I'd prefer if we have enabled MACHINE_FEATURES for all machines that support TF-A and then used `COMBINED_FEATURES` in order to determine whether to use a normal boot flow or a TF-A based one.

ndechesne replied:
I like the COMBINED_FEATURES. 
We will then need to decide if we have an 'open boot' distro ;)

b49020 commented on .github/actions/compile/action.yml:
```diff
@@ -10,6 +10,8 @@ inputs:
     required: true
   kernel_dirname:
     required: true
+  firmware_dirname:
```
Yeah, but we already have different kernel versions built. I think similarly we can support open vs closed firmware builds. However, adding new machine for just open boot firmware will just make all the supported CI artifacts 2x the current number. So, I think current approach in this PR is more scalable.

b49020 replied:
> I don't understand why it would double the artifacts.

I mean if we add a new machine here then all the different distro and kernel versions will have to be built twice.

> Perhaps we can combine the kernel and firmware into a 'variant' attribute, so that we don't keep adding more and more in there?

I am mostly interested to run open boot firmware builds using qcom-next kernel since for fully feature complete we will be contributing patches in kernel for OP-TEE support to enable all the peripherals. So, let me know how you would like to see the variants here.

b49020 commented on conf/machine/qcs6490-rb3gen2-core-kit.conf:
```diff
@@ -13,9 +13,14 @@ KERNEL_DEVICETREE ?= " \
 MACHINE_ESSENTIAL_EXTRA_RRECOMMENDS += " \
     packagegroup-rb3gen2-firmware \
     packagegroup-rb3gen2-hexagon-dsp-binaries \
+    packagegroup-optee \
```
I did that initially, but yocto-check-layers complained about "packagegroup-optee" as an independent entity since a machine compatible is required to build `optee-test` and others.

lumag commented on conf/layer.conf:
```diff
@@ -9,7 +9,7 @@ BBFILE_COLLECTIONS += "qcom"
 BBFILE_PRIORITY_qcom = "6"
 
-LAYERDEPENDS_qcom = "core"
+LAYERDEPENDS_qcom = "core meta-arm"
```
Well... If OP-TEE is not a required feature, this should be handled via dynamic layers rather than layer dependency

lumag replied:
> > At this point, the basic functionality is to use the blobs.
> 
> Is that what `meta-qcom` as an open source BSP layer would enforce its users to use by default even if the open boot stack becomes a viable option in future?
>
> > but I don't want to enforce that dependency on everybody.
> 
> But you are ready to enforce a practice not at all followed by other open source BSP layers? What extra burden do you think meta-arm brings which isn't there for say meta-ti layer?

At this point OP-TEE & TF-A - based stack:
- is not enabled by default on any of the platforms
- doesn't work out of the box outside of QTI
- is provided for a single SoC

Having all of that in mind, as I wrote, I'd welcome the introduction of TF-A & OP-TEE based boot stack (and documenting that in the README), but I don't want to make everybody pull in that dependency. When it's decided that the open boot stack is a default option for any of the targets, we can flip it.

b49020 replied:
> is not enabled by default on any of the platforms

What do you mean by default here? One that is enforced by Qualcomm for every OEM? Or rather we give OEMs the choice to select among open vs closed boot firmware stacks. There are already various IoT and Auto OEMs asking for the open boot firmware stack.

> doesn't work out of the box outside of QTI

As I said this is being worked upon to enable it outside QTI. Also, the next step for me is to add support for upstream edk2 whose recipes are also currently provided by meta-arm which will work out of the box as of now.

> is provided for a single SoC

As I said in the PR itself, the plan is to support Kodiak and Lemans initially, but it will be scalable to other SoCs too with less effort.

> but I don't want to make everybody pull in that dependency.

Sooner or later, we need to add that dependency as otherwise how can we enable OEMs to build open boot stack in the same build environment.

> When it's decided that the open boot stack is a default option for any of the targets, we can flip it.

As I said above, we should really be offering a choice for OEMs to build open boot stacks. And I don't understand when you would feel it's the right time to flip. You need to really understand; the open boot firmware stack is an essential feature of any open source Yocto BSP layer which people expect to be available by default.

lumag commented on recipes-bsp/trusted-firmware-a/trusted-firmware-a-qcom.inc:
```diff
@@ -0,0 +1,22 @@
+SRCBRANCH = "qcom-next"
```
inline

lumag commented on recipes-bsp/trusted-firmware-a/trusted-firmware-a-qcom.inc:
```diff
@@ -0,0 +1,22 @@
+DEPENDS += "qtestsign-native"
+
+MACHINE_TFA_QCOM_REQUIRE ?= ""
+MACHINE_TFA_QCOM_REQUIRE:qcm6490 = "trusted-firmware-a-qcm6490.inc"
```
Will it work on other qcm6490-based machines (e.g. IDP, Tachyon or RubkPi)?

lumag commented on recipes-bsp/trusted-firmware-a/trusted-firmware-a-qcm6490.inc:
```diff
@@ -1,9 +1,13 @@
+TFA_SPD = "opteed"
 EXTRA_OEMAKE:append = " \
     QTISECLIB_PATH=${UNPACKDIR}/libqtisec.a \
+    BL32=${RECIPE_SYSROOT}/${nonarch_base_libdir}/firmware/tee-raw.bin \
```
If you change the order of the commits, you won't have to change the file that you have just added.

b49020 replied:
This commit adds OP-TEE support to the TF-A recipe added by previous commit. So, I am not sure what do you meant by reordering commits here.

lumag replied:
- Add OP-TEE as a recipe
- Add TF-A, which uses OP-TEE among other dependencies

lumag commented on recipes-security/optee/optee-os-qcm6490.inc:
```diff
@@ -0,0 +1,6 @@
+
+require optee-os-qcom.inc
+
+COMPATIBLE_MACHINE = "qcm6490"
```
Same comment: I'd prefer if we can build binaries for all supported machines

lumag commented on recipes-security/packagegroups/packagegroup-optee.bb:
```diff
@@ -0,0 +1,9 @@
+RRECOMMENDS:${PN} = " \
+    ${@bb.utils.contains('MACHINE_FEATURES', 'optee', 'optee-client', '', d)} \
+    ${@bb.utils.contains('MACHINE_FEATURES', 'optee', 'optee-test', '', d)} \
+    ${@bb.utils.contains('MACHINE_FEATURES', 'optee', 'optee-os-ta', '', d)} \
```
Why do we need three lines here?

lumag commented on recipes-bsp/trusted-firmware-a/trusted-firmware-a-qcom.inc:
```diff
@@ -0,0 +1,22 @@
+SRCBRANCH = "qcom-next"
+SRC_URI = "git://github.com/qualcomm-linux/trusted-firmware-a.git;protocol=https;name=tfa;branch=${SRCBRANCH}"
+SRCREV_tfa = "164d443c8e92c18fd2cca03661dd1a2b9ac37848"
```
This breaks the idea of the orignial TF-A recipes. Inside meta-arm there are 4 different versions of TF-A (2.10.17, 2.12.3, 2.13.0, git aka tip). Machine-specific includes don't change the URI and the SRCREV. In other words: if the user sets `PREFERRED_VERSION_trusted-firmware-a`, it is expected to get _that_ TF-A version, not some random machine-specific version

lumag commented on recipes-security/optee/optee-os-qcom.inc:
```diff
@@ -0,0 +1,3 @@
+SRCBRANCH = "qcom-next"
+SRC_URI = "git://github.com/qualcomm-linux/optee_os.git;protocol=https;name=optee;branch=${SRCBRANCH}"
+SRCREV_optee = "a427f12bc60e74ccdd2ae771b50ed237397a0782"
```
The same question. We rewrite SRC_URI and SRCREV here. It breaks user assumptions.

lumag commented on conf/machine/qcs6490-rb3gen2-core-kit.conf:
```diff
@@ -13,9 +13,14 @@ KERNEL_DEVICETREE ?= " \
 MACHINE_ESSENTIAL_EXTRA_RRECOMMENDS += " \
     packagegroup-rb3gen2-firmware \
     packagegroup-rb3gen2-hexagon-dsp-binaries \
+    packagegroup-optee \
```
Please keep the list sorted

lumag commented on conf/machine/qcs6490-rb3gen2-core-kit.conf:
```diff
@@ -13,9 +13,14 @@ KERNEL_DEVICETREE ?= " \
 
+EXTRA_IMAGEDEPENDS += "${@bb.utils.contains('MACHINE_FEATURES', 'tfa', 'trusted-firmware-a', '', d)}"
+PREFERRED_PROVIDER_virtual/bootloader = "${@bb.utils.contains('MACHINE_FEATURES','u-boot','u-boot','',d)}"
+TFA_UBOOT:pn-trusted-firmware-a = "${@bb.utils.contains('MACHINE_FEATURES','u-boot','1','0',d)}"
```
It feels strange at a first glance that machine configuration (which should also define MACHINE_FEATURES) also depends on them. Please add a comment here, how is it expected to be enabled.

ndechesne commented on .github/workflows/build-yocto.yml:
```diff
@@ -144,6 +145,14 @@ jobs:
+            firmware:
+                type: additional
+                dirname: "+open-boot-firmware"
+                yamlfile: ":ci/open-boot-firmware.yml"
```
the yamlfile is unused. does not sound right.

b49020 replied:
Ah, I will fix it.

lumag commented on recipes-bsp/trusted-firmware-a/trusted-firmware-a-qcm6490.inc:
```diff
@@ -0,0 +1,9 @@
+SRC_URI += "https://github.com/coreboot/qc_blobs/raw/refs/heads/master/sc7280/qtiseclib/libqtisec.a;name=qtiseclib"
```
Still the licence should be included in SBOM, etc., so it should be properly annotated.
Also we have to comply to the licensing terms. E.g.:

> (iii) redistribution of the Redistributable Binary Code must include the .txt file setting forth the terms and condition of this Agreement;

b49020 commented on .github/actions/compile/action.yml:
```diff
@@ -10,6 +10,8 @@ inputs:
     required: true
```
> If it requires extra signing step, it is of very limited use.

Sure, it's a temporary extra step until we enable the OEM only signed TZ image. But still, it's a good enough first step for people working within Qualcomm to collaborate while making the open boot firmware stack a reality on Qcom silicon.

nrclark commented:
Just wanted to leave a comment here - I'm very glad to see this develop. GM and many of Qualcomm's other customers would strongly prefer a `TF-A`+`OP-TEE`+`U-boot` stack over the current proprietary mess. You would make my day if you could get the automotive group to adopt it for Lemans and Nord.

One extra item to note: if you don't have one already, you should consider setting up a uboot config that can be loaded with Sahara. That would allow customers to get away from Firehose and flatbuilds in favor of more cross-platform solutions inside of uboot.

quaresmajose commented:
Yes, this work tree can be broken because all the others have been updated in the meantime. The meta-yocto 
is the one which provides the linux-yocto.

Please rebase and pick the latest tip of all layers with:`kas build --update ci/qcs6490-rb3gen2-core-kit.yml:ci/open-boot-firmware.yml`.

The other way arround is locking the revision of all the layers when the PR runs last time and extract the [kas-lock.zip](https://github.com/qualcomm-linux/meta-qcom/actions/runs/19125459442/artifacts/4483017157) to `ci/base.lock.yml`. Then run `kas build ci/qcs6490-rb3gen2-core-kit.yml:ci/open-boot-firmware.yml`. Don't forget to remove the `ci/base.lock.yml` when it is no longer necessary.

ldts commented:
> Yes, this work tree can be broken because all the others have been updated in the meantime. The meta-yocto is the one which provides the linux-yocto.
> 
> Please rebase and pick the latest tip of all layers with:`kas build --update ci/qcs6490-rb3gen2-core-kit.yml:ci/open-boot-firmware.yml`.
> 
> The other way arround is locking the revision of all the layers when the PR runs last time and extract the [kas-lock.zip](https://github.com/qualcomm-linux/meta-qcom/actions/runs/19125459442/artifacts/4483017157) to `ci/base.lock.yml`. Then run `kas build ci/qcs6490-rb3gen2-core-kit.yml:ci/open-boot-firmware.yml`. Don't forget to remove the `ci/base.lock.yml` when it is no longer necessary.

thanks @quaresmajose ! yeah I forgot it is a moving target. I am chasing an issue with a secure watchdog that is being affected by the kernel boot sequence since 6.18-rc4 (still not sure what the heck is going on but the 6 second barking period programmed by TF-A is resulting in >22 during linux boot leading to a watchdog bite and reset). anyway thanks for the options!

quaresmajose commented:
There is a proposal to move OP-TEE from meta-arm to oe-core on the mailing list [[RFC] Moving OP-TEE support from meta-arm to oe-core
](https://lists.yoctoproject.org/g/yocto/message/66178)

b49020 committed:
recipes-bsp: Add trusted-firmware-a support for RB3Gen2

Add RB3Gen2 platform specific support leveraging base trusted-firmware-a
recipes from meta-arm. It allows to generate unsigned bl2.elf and test signed
fip.elf as per TF-A documentation here [1].

[1] https://github.com/qualcomm-linux/trusted-firmware-a/blob/qcom-next/docs/plat/qti/rb3gen2.rst

Signed-off-by: Sumit Garg <sumit.garg@oss.qualcomm.com>

b49020 committed:
layer.conf: add meta-arm support via dynamic-layers

meta-arm provides the base recipes for open boot firmware on Arm based
platforms. So lets add meta-arm as an optional dependency via dynamic
layers such that Qcom targets can start using open boot firmware stack.

Signed-off-by: Sumit Garg <sumit.garg@oss.qualcomm.com>

b49020 committed:
conf: manchine: Add new machine rb3gen2-core-kit-open-fw

Add support for new RB3Gen2 machine type with support for open boot firmware based
on TF-A, OP-TEE and U-Boot.

Right now the build generates 2 firmware payloads as bl2.elf (unsigned) and fip.elf
(test signed using qtestsign). It is required to sign bl2.elf with QTI signature
using sectools but in future the plan is to drop QTI signature requirement with an
updated release of XBL/XBL-SEC.

Once signing is done, one need to update following binaries in qcomflash tarball
and then proceed with QDL flashing:
- tz.mbn -> bl2.mbn
- uefi.elf -> fip.elf

Signed-off-by: Sumit Garg <sumit.garg@oss.qualcomm.com>

b49020 committed:
ci: add KAS configuration files for rb3gen2-core-kit-open-fw

Add KAS configuration file for RB3Gen2 with open boot firmware machine type.
With open boot firmware stack, UEFI boots up Linux in EL2 mode by default and
that is the only supported option as of now compliant with Arm SystemReady
standards. Hence, Linux with KVM enabled is the only supported configuration.

Signed-off-by: Sumit Garg <sumit.garg@oss.qualcomm.com>

ndechesne commented on conf/layer.conf:
```diff
@@ -22,6 +22,8 @@ BBFILES_DYNAMIC += " \
     networking-layer:${LAYERDIR}/dynamic-layers/networking-layer/*/*/*.bbappend \
     selinux:${LAYERDIR}/dynamic-layers/selinux/*/*/*.bb \
     selinux:${LAYERDIR}/dynamic-layers/selinux/*/*/*.bbappend \
+    meta-arm:${LAYERDIR}/dynamic-layers/meta-arm/*/*/*.bb \
```
unless there is a reason (if so, let's document it), please keep list sorted.

lumag commented on dynamic-layers/meta-arm/recipes-bsp/trusted-firmware-a/trusted-firmware-a-qcom_git.bb:
```diff
@@ -0,0 +1,16 @@
+require recipes-bsp/trusted-firmware-a/trusted-firmware-a.inc
+
+DEPENDS += "qtestsign-native"
```
This should be a part of the qcm6490 include.

lumag reviewed (changes requested):
Overall looks great.

ricardosalveti commented on dynamic-layers/meta-arm/recipes-security/optee/optee-qcm6490.inc:
```diff
@@ -0,0 +1,4 @@
+# RB3Gen2 specific OP-TEE support
+
+COMPATIBLE_MACHINE = "qcm6490"
+OPTEEMACHINE = "qcom-kodiak"
```
I would probably just make this include generic to qcom and define these variables for qcm6490 only.

b49020 replied:
Sounds reasonable, will do that.

ricardosalveti commented on dynamic-layers/meta-arm/recipes-security/packagegroups/packagegroup-optee.bb:
```diff
@@ -0,0 +1,7 @@
+
+RRECOMMENDS:${PN} = " \
+    ${@bb.utils.contains('MACHINE_FEATURES', 'optee', 'optee-os-tadevkit optee-client optee-test', '', d)} \
+"
```
Please break this packagegroup into a separated commit.

b49020 commented on ci/rb3gen2-core-kit-open-fw.yml:
```diff
@@ -0,0 +1,19 @@
+
+local_conf_header:
+  optee: |
+    MACHINE_EXTRA_RRECOMMENDS += "packagegroup-optee"
```
Yocto check layers doesn't allow me to put this as part of machine config as all the OP-TEE components can't be provided by `meta-qcom` layer since `meta-arm` support comes via dynamic layers.

b49020 commented on dynamic-layers/meta-arm/recipes-bsp/trusted-firmware-a/trusted-firmware-a-qcm6490.inc:
```diff
@@ -0,0 +1,31 @@
+SRC_URI += " \
+    https://github.com/coreboot/qc_blobs/raw/refs/heads/master/sc7280/qtiseclib/libqtisec.a;name=qtiseclib \
+    https://github.com/coreboot/qc_blobs/raw/refs/heads/master/sc7280/qtiseclib/LICENSE;name=qtiseclib_lic \
```
I don't see an exact match with qcom or qcom-2, there are minor wording differences. Let me rather follow Dmitry's suggestion with: `LICENSE += "LICENSE.qcom-2"` and also keep this additional license file explicitly.

b49020 commented on dynamic-layers/meta-arm/recipes-security/optee/optee-os-qcom_git.bb:
```diff
@@ -0,0 +1,11 @@
+SRCREV_optee = "c2b0684fcd89929976a8726e6e3af922b48dd2c7"
+
+MACHINE_OPTEE_REQUIRE ?= ""
+MACHINE_OPTEE_REQUIRE:qcm6490 = "optee-qcm6490.inc"
```
Exploring further I see TF-A recipes allows Qemu machine support via `trusted-firmware-a_%.bbappend` which won't apply to qcom recipe variant. So I would just by default include the *.inc file.

b49020 committed:
recipes-security: Add OP-TEE recipes override for RB3Gen2

Add OP-TEE recipes override for RB3Gen2 leveraging the base recipes from
meta-arm.

Signed-off-by: Sumit Garg <sumit.garg@oss.qualcomm.com>

b49020 committed:
recipes-security: Add OP-TEE packagegroup recipe

Add OP-TEE packagegroup to add all the OP-TEE components if machine
supports OP-TEE feature.

Signed-off-by: Sumit Garg <sumit.garg@oss.qualcomm.com>

koenkooi commented:
> @koenkooi I needed an initramfs change to break a circular dependency reported by CI, here: [3bdcf07](https://github.com/qualcomm-linux/meta-qcom/commit/3bdcf07275cd1e5cf2e7a7b5a9a2908109f1e4c2). Not sure why this isn't a problem with initramfs recipes provided by OE core. Or is meta-qcom initramfs recipe special in some sense here?

Our recipe is functionally the same as `openembedded-core/meta/recipes-core/images/core-image-initramfs-boot.bb`, I suspect the difference is in that we use both u-boot and TF-A which change the dependencies. I need to have a better look at those to judge if `IMAGE_DEPENDS = ""` is the proper solution.

b49020 commented:
@koenkooi 

> I suspect the difference is in that we use both u-boot and TF-A which change the dependencies. I need to have a better look at those to judge if IMAGE_DEPENDS = "" is the proper solution.

Okay I have switched to your suggestion instead, let me know if that's fine.

ricardosalveti commented on .github/workflows/build-yocto.yml:
```diff
@@ -219,6 +219,14 @@ jobs:
+          - machine: rb3gen2-core-kit-open-fw
+            distro:
+                name: qcom-distro-kvm
+                yamlfile: ':ci/qcom-distro-kvm.yml'
```
Mind extending the commit message explaining why qcom-distro-kvm is used here?

ricardosalveti commented:
> Okay I have switched to your suggestion instead, let me know if that's fine.

Which build failed because of this? I want to reproduce locally.

b49020 commented:
> > Okay I have switched to your suggestion instead, let me know if that's fine.
> 
> Which build failed because of this? I want to reproduce locally.

Just try following, it will report the circular dependency:

```
$ kas build ../meta-qcom/ci/rb3gen2-core-kit-open-fw.yml:../meta-qcom/ci/qcom-distro-kvm.yml
```

ricardosalveti commented:
> That I plan to do as a follow up PR. Right now I need to figure out how to enable OEM signing for BL2 TZ image. We have been using sectools internally but there isn't a public release of profiles yet. With qtestsign, I need to see if that can support OEM only signing for TZ image.

Sure, sounds good.

CodeChenL commented:
Hello everyone, I am an operating systems engineer from Radxa. We are currently looking to begin using Open Boot firmware on our Qualcomm products such as [SBCs](https://docs.radxa.com/dragon/q6a), so our users and customers can customize their own boot firmware and, together with more people, make them better.
But when I tried to use this boot chain, it entered an infinite reboot when booting to BL31. Logs:
https://gist.github.com/CodeChenL/6b48eeaf33fc5f0a5738cde699c09941#file-gistfile1-txt
After my debugging I found it might be stuck at the final cpucp-related call: https://gist.github.com/CodeChenL/6b48eeaf33fc5f0a5738cde699c09941?permalink_comment_id=6101921#gistcomment-6101921
I would be very grateful if anyone in this thread could give me some help

b49020 commented:
Thanks @CodeChenL for your report, can you rather create an issue under Qcom TF-A staging tree (https://github.com/qualcomm-linux/trusted-firmware-a/issues) here such that the status can be tracked?

CodeChenL commented:
> Thanks @CodeChenL for your report, can you rather create an issue under Qcom TF-A staging tree (https://github.com/qualcomm-linux/trusted-firmware-a/issues) here such that the status can be tracked?

Thank you for your reply. After our testing and discussion, we have decided to shift our development focus to upstream EDK2. I have already ported an MMC driver and successfully booted the system using it. Thank you all for your work; it's truly great and has set a good precedent for us

b49020 commented:
@CodeChenL I assume you would have used following command to sign TF-A BL2:

```
sectools secure-image bl2.elf \
    --outfile bl2.mbn \
    --image-id TZ-TEE \
    --security-profile kodiak_tz_security_profile.xml \
    --sign --signing-mode TEST
```
actually I am able to reproduce the issue you are seeing here on RB3Gen2. Looks like OEM only signing feature in XBL_SEC is leading to this issue. The QTI signed BL2 works fine though. We will keep digging into this issue, I have filed one here: https://github.com/qualcomm-linux/trusted-firmware-a/issues/19 so you can keep track of status there.

> Thank you for your reply. After our testing and discussion, we have decided to shift our development focus to upstream EDK2. I have already ported an MMC driver and successfully booted the system using it. Thank you all for your work; it's truly great and has set a good precedent for us

PRs are always welcome to upstream edk2 project extending Qcom platforms support. I hope the open boot stack effort is valuable to your customers/developers.

CodeChenL commented:
> @CodeChenL I assume you would have used following command to sign TF-A BL2:
> 
> ```
> sectools secure-image bl2.elf \
>     --outfile bl2.mbn \
>     --image-id TZ-TEE \
>     --security-profile kodiak_tz_security_profile.xml \
>     --sign --signing-mode TEST
> ```
> 
> actually I am able to reproduce the issue you are seeing here on RB3Gen2. Looks like OEM only signing feature in XBL_SEC is leading to this issue. The QTI signed BL2 works fine though. We will keep digging into this issue, I have filed one here: [qualcomm-linux/trusted-firmware-a#19](https://github.com/qualcomm-linux/trusted-firmware-a/issues/19) so you can keep track of status there.
> 
> > Thank you for your reply. After our testing and discussion, we have decided to shift our development focus to upstream EDK2. I have already ported an MMC driver and successfully booted the system using it. Thank you all for your work; it's truly great and has set a good precedent for us
> 
> PRs are always welcome to upstream edk2 project extending Qcom platforms support. I hope the open boot stack effort is valuable to your customers/developers.

When i complete the basic functionality and pass our internal team's code quality review, I will submit them to the upstream EDK2
